plugins/package-managers/bazel/src/main/kotlin/CompositeBazelModuleRegistryService.kt

```diff
@@ -0,0 +1,102 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.plugins.packagemanagers.bazel
+
+import java.io.File
+
+import org.apache.logging.log4j.kotlin.logger
+
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.BazelModuleRegistryService
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.LocalBazelModuleRegistryService
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.ModuleMetadata
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.ModuleSourceInfo
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.RemoteBazelModuleRegistryService
+
+/**
+ * A composite Bazel module registry service that aggregates multiple [BazelModuleRegistryService] instances and
+ * delegates the package lookup to the corresponding registry. The first registry that contains the package is used.
+ * [packagesPerRegistry] is a map of [BazelModuleRegistryService] instances to the list of packages they are
+ * responsible for.
+ */
+class CompositeBazelModuleRegistryService(
+    private val packagesPerRegistry: Map<BazelModuleRegistryService, List<String>>
+) : BazelModuleRegistryService {
+    companion object {
+        // A regular expression to extract the server and package parts of a registry URL.
+        internal val URL_REGEX = "^(?<server>.*/)modules/(?<package>[^/]+)/[^/]+/source\\.json$".toRegex()
+
+        /**
+         * Create a Composite Bazel Module Registry client instance.
+         */
+        fun create(urls: Set<String>, projectDir: File): CompositeBazelModuleRegistryService {
+            val unknownKey = "UNKNOWN"
+
+            val packagesPerServer = urls.filter { it.endsWith("source.json") }.map {
+                val groups = URL_REGEX.matchEntire(it)?.groups
+
+                val key = groups?.get("server")?.value ?: unknownKey
+                val value = groups?.get("package")?.value ?: it
+
+                key to value
+            }.groupByTo(mutableMapOf(), { it.first }) { it.second }
+
+            packagesPerServer.remove(unknownKey)?.let {
+                logger.warn {
+                    "The following URLs cannot be mapped to a server root:"
+                }
+
+                it.forEach { logger.warn { it } }
+            }
+
+            val packagesPerRegistry = packagesPerServer.mapKeys { entry ->
+                val url = entry.key
+
+                if (url.startsWith("file://")) {
+                    val localRegistryURL = url.replace("%workspace%", projectDir.absolutePath)
+                    logger.info {
+                        "Using local Bazel module registry at '$localRegistryURL'."
+                    }
```

Comment:
There is no analog log statement for the remote case, so I believe for the local case it is not needed either and can be removed. (It could be made consistent also the other way round.